KAFKA-15276: Implement partition assignment reconciliation

kirktrue opened the pull request:
Added logic for reconciling partition assignment between the target assignment provided by the group coordinator and the current assignment as specified in the `SubscriptionState`.

This refactors the `ConsumerRebalanceListener` code from `ConsumerCoordinator` for reuse in both places.

kirktrue committed:
KAFKA-15276: Implement partition assignment reconciliation

Added logic for reconciling partition assignment between the target
assignment provided by the group coordinator and the current assignment
as specified in the SubscriptionState.

This refactors the ConsumerRebalanceListener code from
ConsumerCoordinator for reuse in both places.

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/MemberAssignmentReconciler.java:
```diff
@@ -0,0 +1,291 @@
+                                      SubscriptionState subscriptions,
+                                      ConsumerMetadata metadata,
+                                      BlockingQueue<BackgroundEvent> backgroundEventQueue) {
+        this.log = logContext.logger(getClass());
```
I wonder what's the convention for the use of `this`.  I got pointed out several times by jun for not using it.

kirktrue replied:
Use of `this` when assigning instance variables is pretty common. Use of `this` to refer to the current object under other situations isn't strictly essential.

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/MemberAssignmentReconciler.java:
```diff
@@ -0,0 +1,291 @@
+        }
+    }
+
+    static SortedSet<TopicPartition> sortPartitions(Set<TopicPartition> topicPartitions) {
```
Do we actually need to sort the partition? I've seen that several times in the current code, but I never understand to reason for sorting.

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/MemberAssignmentReconciler.java:
```diff
@@ -0,0 +1,291 @@
+ * <em>flattening</em> the respective entries into two sets of {@link org.apache.kafka.common.TopicPartition partitons}
+ * which are then compared using basic {@link Set} comparisons.
+ */
+public class MemberAssignmentReconciler {
```
One missing piece would be the heartbeat - the heartbeat needs to be sent out upon revoking partitions. I think we can leave a note here or implement this in this PR.

kirktrue replied:
The `revoke()` and `assign()` methods return a `ReconciliationResult` to let the caller know the operation completed. In that case, the caller would work with the `HeartbeatRequestManager` so that it includes the updated partition set in its next request.

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java:
```diff
@@ -237,4 +253,64 @@ public void updateAssignment(ConsumerGroupHeartbeatResponseData.Assignment assig
+                // let the coordinator know that they've been removed locally.
+                return;
+            } else if (result == ReconciliationResult.EXPIRED) {
+                // TODO: what do we do here?
```
This is the description from kip-848 
```
The chosen member is expected to complete the assignment process within the rebalance timeout. The time on the coordinator side starts ticking when the member is notified. If the process is not completed within the rebalance timeout, the group coordinator picks up another member to run the assignment. Note that the previous chosen member is not fenced here because the fencing is only done based on the session.
```

It seems that you should send a heartbeat to the GC and get a fenced exception, which results in revoking all partitions.

@dajac can provide a more accurate answer here.

lianetm replied:
I may be missing something, but I don't quite get why we need to time this on the client side. It is the server the one that handles the rebalanceTimeouts. 

The server will start the timer when sending an assignment/revocation to a member, and if it does not receive the ack in time, it will fence the member. From the client point of view, when it completes processing the assignment/revocation, it would notice it has been fenced (state is FENCED), and it won't send the ack to the assignment (because it is not part of the group anymore). It would have to invoke the onPartitionsLost and rejoin the group I expect.

kirktrue replied:
Should I remove the `Timer` and expose an `onFenced()` method that the heartbeat invokes? Then the `onFenced()` implementation would call `onPartitionsLost()`, right?

lianetm replied:
I would say we should aim for keeping only state updates and member info in the MembershipMgr, not the logic for triggering the `onPartitionsLost`. It could be maybe the assignment reconciler that, when it finishes processing, triggers a HB or a partitionsLost depending on the state (if state is still reconciling and it finished processing ok we just need to HB, if state is fenced then we don't HB. We trigger onPartitionsLost and then trigger HB to rejoin). 
Seems all like interaction between the AssginemntReconciler and the HB request, only checking the state to get the current state and determine what to do (and eventually updating the state with the outcome)

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java:
```diff
@@ -237,4 +253,64 @@ public void updateAssignment(ConsumerGroupHeartbeatResponseData.Assignment assig
+            remainingAssignmentTime.update();
+
+            if (result == ReconciliationResult.COMPLETED) {
+                // If we've revoked one or more partitions, we need to send an acknowledgement request ASAP to
```
I'm having a bit of dilemma of how to correct handling the heartbeat timer reset here - now, we could have this and the heartbeat request manager to hold on to each other, and invokes "onAssignmentRevocation" to expire the heartbeat timer and ensure the next poll will have the heartbeat sent out.  

Another way to do it, which could be better, is to have the heartbeat check on a state here to determine if a heartbeat needs to be sent out immediately.

lianetm replied:
I get your dilemma, and I think the root cause relates to my comment above, and this whole reconcile logic needs to be driven by the HB, not the state. What about we define the components such that:

- HeartbeatReqMgr: responsible for preparing and timing HB request, and handling responses. Handling responses involves:
    - update shared stated based on HB responses (calling membershipMgr)
    - trigger assignment processing (calling AssignmentReconciler), which will directly affect the timing of the HB requests.
- AssignmentReconciler: responsible for processing an assignment (invoke callbacks & update shared state)
- MembershipMgr: responsible for keeping member info and state. This is mainly a shared state between HB manager and AssignmentReconciler, nothing more. 

The HeartbeatReqMgr is the central player, which makes sense because everything is based on the HB requests, And it relies on the AssignmentReconciler (to delegate assignment processing), and on the MembershipMgr (just to keep shared info in one place).

(Let's have a chat in our sync)

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/CompletableBackgroundEvent.java:
```diff
@@ -0,0 +1,86 @@
+
+    @Override
+    public String toString() {
+        return getClass().getSimpleName() + "{" +
```
we should use ( ) I think @dajac pointed out one day.

kirktrue replied:
Unfortunately it seems pretty inconsistent. `MetadataCache`, as the first example I could find, uses the more "standard" Java brackets 🤷‍♂️

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java:
```diff
@@ -237,4 +253,64 @@ public void updateAssignment(ConsumerGroupHeartbeatResponseData.Assignment assig
+     * TODO: Is SubscriptionState sufficient to build the next heartbeat request?
+     * TODO: Where do we need to call ConsumerRebalanceListener.onPartitionsLost()?
+     */
+    void reconcile() {
```
I wonder if the timer should be updated here. We should be checking the reconciliation results on every poll no?

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java:
```diff
@@ -237,4 +253,64 @@ public void updateAssignment(ConsumerGroupHeartbeatResponseData.Assignment assig
+    void reconcile() {
+        transitionTo(MemberState.RECONCILING);
+
+        Timer remainingAssignmentTime = time.timer(10000);
```
You should be getting the rebalance timeout from the max poll interval configuration

```
The rebalance timeout is provided by the member when it joins the group. It is basically the max poll interval configured on the client side
```

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/MemberAssignmentReconciler.java:
```diff
@@ -0,0 +1,291 @@
+                targetPartitions(metadata, assignment)));
+
+        // TODO: should this be here? Do we expect the above to do anything besides move things around in memory?
+        timer.update();
```
actually - you already are updating the timer no? It doesn't seem necessary here.

philipnee commented:
Hi @kirktrue - This is a really nice PR. I left some comments but overall the strategy seems right.

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/MemberAssignmentReconciler.java:
```diff
@@ -0,0 +1,291 @@
+        event.future().whenComplete((result, error) -> {
+            if (error != null) {
+                log.warn("An error occurred when invoking {} with the following partitions: {}", operation.methodName, diffPartitions, error);
+                // TODO: should we proceed or abort?
```
The old consumer completely fails (abort the poll loop) with a KafkaException when there is a failure in any of the callbacks (exception starting [here](https://github.com/apache/kafka/blob/9e5ca8416de6011fa8c0d63430e9b89fe751ee2b/clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java#L485))

kirktrue replied:
This is tricky because it doesn't fail immediately. It waits until _after_ the assignment step. So now we need to keep track of that exception across multiple heartbeats and application<->background thread hops 😢

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/MemberAssignmentReconciler.java:
```diff
@@ -0,0 +1,291 @@
+            assignmentModifier.modify(newAssignment, diffPartitions);
+
+            // And finally... assign the new set of partitions!
+            subscriptions.assignFromSubscribed(newAssignment);
```
I would add a comment pointing out that the subscription can only be updated after the callbacks are executed so we make sure that user will be able to commit offsets in the callbacks on revocation. 

We have the same behaviour for assignment/revocation but it's truly the revocation+commit what forces this dependency between the callback exec and making the subscription effective (and start fetching)

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java:
```diff
@@ -237,4 +253,64 @@ public void updateAssignment(ConsumerGroupHeartbeatResponseData.Assignment assig
+
+        // First, we need to determine if any partitions need to be revoked.
+        {
+            ReconciliationResult result = reconciler.revoke(targetAssignment, remainingAssignmentTime);
```
I was expecting it was going to be the heartbeat manager the one calling this kind of logic . The state machine shouldn't be involved in anything other than keeping member info and state, and it already transitions to RECONCILE the moment the heartbeat req manager calls the `updateState` with a successful heartbeat response.

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java:
```diff
@@ -237,4 +253,64 @@ public void updateAssignment(ConsumerGroupHeartbeatResponseData.Assignment assig
+     *
+     * TODO: What are the the state changes here?
+     * TODO: Is SubscriptionState sufficient to build the next heartbeat request?
+     * TODO: Where do we need to call ConsumerRebalanceListener.onPartitionsLost()?
```
I don't think invoking the `onPartitionsLost` should be driven by the state machine but just for the record, we need to call `onPartitionsLost` whenever the member is left out of the group and it has an assignment. The basic case is when we receive a HB response with a fencing error. 
This needs more thought, mainly around what other cases would lead to a member out of a group while still holding an assignment.

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java:
```diff
@@ -237,4 +253,64 @@ public void updateAssignment(ConsumerGroupHeartbeatResponseData.Assignment assig
+            remainingAssignmentTime.update();
+
+            if (result == ReconciliationResult.COMPLETED) {
+                // If we've assigned one or more partitions, we need to send an acknowledgement request ASAP to
```
This should be interaction between the AssignmentReconciler and the HB manager I would expect. The state machine should only care about updating the state to RECONCILING when the assignment process starts, and to STABLE/FAILED when it completes.

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AssignmentReconciler.java:
```diff
@@ -44,21 +45,46 @@
+ *
+ * <ol>
+ *     <li>Calculating partitions to revoke</li>
+ *     <li>Invoking {@link ConsumerRebalanceListener#onPartitionsRevoked(Collection)}</li>
```
We need to commit offsets before revoking partitions if auto commit is enabled, I expect that will be here right? (interacting with the commitRequest manager)

kirktrue replied:
Yes, that has been split off into KAFKA-15573.

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AssignmentReconciler.java:
```diff
@@ -0,0 +1,355 @@
+            SortedSet<TopicPartition> newAssignment = assignmentGenerator.generate(diffPartitions);
+
+            // And finally... assign the new set of partitions!
+            subscriptions.assignFromSubscribed(newAssignment);
```
Just for the record, this is the link between this component and the fetching logic, based on the existing subscription state. I expect that after this line, the next iteration of the poll loop will start fetching from the new partitions (if not paused and if they have a valid position)

kirktrue replied:
Correct.

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AssignmentReconciler.java:
```diff
@@ -85,122 +111,162 @@ enum ReconciliationResult {
+            // We don't actually need the _result_ of the event, just to know that it's complete.
+            if (inflightCallback.get().future().isDone()) {
+                // This is the happy path--we completed the callback. Clear out our inflight callback first, though.
+                inflightCallback = Optional.empty();
```
What about the case where the callback execution does not complete? We need to somehow clear this I expect. (If the callback does not complete, the broker will eventually kick the member out of the group, the member will rejoin so there should be a new `inflightCallback`)

kirktrue replied:
Is the `HeartbeatRequestManager` going to call `AssignmentReconciler.lose()` to drop the partitions at that point? If so, this is the code at the top of that `lose()` method:

```java
    ReconciliationResult lose() {
        // Clear the inflight callback reference. This is done regardless of if one existed; if there was one it is
        // now abandoned because we're going to "lose" our partitions. This will also allow us to skip the inflight
        // check the other steps take.
        inflightCallback = Optional.empty();

        . . . 
```

Does that seem sufficient, or is more needed?

kirktrue replied:
This is unrelated and will be resolved.

kirktrue commented:
@lianetm @philipnee @dajac we can start reviewing this PR now. I'll keep working on the unit tests in the background.

I resolved a bunch of comments that pertained to a former implementation approach that we abandoned. If any are still relevant, feel free to un-resolve them.

Thanks!

kirktrue committed:
Resolved linter complaints

kirktrue committed:
Moved processRebalanceCallback method from PrototypeAsyncConsumer to ConsumerUtils to allow reuse in AssignmentReconcilerTest

kirktrue committed:
Added more unit tests

kirktrue committed:
Minor refactoring on reconciler unit test

kirktrue committed:
Minor logging and whitespace clean up in AssignmentReconciler

kirktrue committed:
More AssignmentReconcilerTest updates

kirktrue committed:
Reworked design of AssignmentReconciler

API is cleaner
Code executes on the correct threads
Updated unit tests and comments

dajac commented:
@kirktrue Could you briefly explain how the reconciler will be integrated into the rest? I would like to better understand from where `maybeAssign`/`maybeRevoke` will be called. Thanks.

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AssignmentReconciler.java:
```diff
@@ -0,0 +1,328 @@
+ *         {@link SubscriptionState#assignFromSubscribed(Collection) current assignment}
+ *     </li>
+ *     <li>
+ *         Signal to the heartbeat request manager to perform a heartbeat acknowledgement with the group coordinator
```
I don't fully understand this. My mental model for the client side implementation is the following. The client receives a new target assignment. It computes the difference between it and its current assignment. This gives the set of partitions to revoke and the partitions to assign. Then, it does the revocation followed by the assignment of those partitions. Finally, it acknowledges the target assignment. It seems that we handle it slightly differently here. I may have misunderstood it though.

kirktrue replied:
@dajac I have it in my head that the heartbeat request manager sends a heartbeat after revocation, and then again after assignment? i.e. it's a two-step back-and-forth. Should the client wait to send the next heartbeat after assignment is totally complete on the client?

lianetm replied:
My mental model was similar to yours @dajac , just the clarification that I was expecting that the client would acknowledge revocations with a HB before processing any new assignment (in the case of a client that receives partitions to revoke and new partitions in the same HB response) 

Ex. client that is part of a group and owns partition A, from the client point of view:
- receives HB response with assignment [A,B]
- notices new partition B 
   - client performs assignment (callbacks)
   - sends ack as HB with owned topicPartitions [A,B]
- receives HB response with assignment [B,C]
- notices revocation of A and new assignment C. 
   - Client does revocation (callback), **sends ack as HB with owned topicPartitions [B]** to ack the revocation asap
   - Client does assignment of C (callback), sends ack as HB with owned TopicPartitions [B,C]
   
This is my understanding of the expected flow, and my explanation for @kirktrue 's comment. Is it correct?

dajac replied:
> Client does revocation (callback), sends ack as HB with owned topicPartitions [B] to ack the revocation asap

This is incorrect. The client should ack the whole assignment that it has received. The assignment is the unit of work, not the individual partitions.

The reconciliation loop should be as follow:
1. Receive new target
2. Diff target and current
3. Revoke partitions (commit offsets & callback)
4. Assign partitions (commit offsets & callback)
5. Ack the assignment

lianetm replied:
Got it. I was mistaken then, but it's really a simplification on the client side so sounds good to me.

lianetm replied:
Just to recap, it's fair to say then that the reconciliation process when there are revocations and new assignments involves a single interaction between client-server (HB req + process revocation/assignment + HB resp), vs. the double rebalance that takes places with the old protocol in the same scenario, correct?

kirktrue replied:
The code has been changed to perform callbacks for both revoke and assign in one shot, with no heartbeats between them, as discussed above.

kirktrue committed:
Updates to documentation and event class names

kirktrue committed:
Merge remote-tracking branch 'upstream/trunk' into KAFKA-15276-partition-reconciliation

kirktrue committed:
Substantial overhaul to the assignment reconciliation code based on a single heartbeat vs. two

kirktrue committed:
Update ConsumerRebalanceListenerInvoker.java

kirktrue committed:
Added ConsumerCoordinatorMetrics to make ConsumerRebalanceListenerInvoker cleaner

kirktrue committed:
Clean up of naming in AssignmentReconciler internal methods

kirktrue committed:
Removed CompletableBackgroundEvent as it violates the inter-thread tenets

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java:
```diff
@@ -163,6 +176,32 @@ public NetworkClientDelegate.PollResult poll(long currentTimeMs) {
+            //       return?
+        }
+
+        assignmentReconciler.reconciliationCallbacksInvoked(revokedPartitions, assignedPartitions);
```
Need to update state in the membership manager

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java:
```diff
@@ -163,6 +176,32 @@ public NetworkClientDelegate.PollResult poll(long currentTimeMs) {
+            //       return?
+        }
+
+        assignmentReconciler.lostCallbackInvoked();
```
Need to update state in the membership manager

kirktrue committed:
Updates to move the AssignmentReconciler inside the MembershipManager

kirktrue committed:
Update HeartbeatRequestManager.java

Removed the version of the resetCall method that is no longer used

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java:
```diff
@@ -135,6 +146,7 @@ public void updateState(ConsumerGroupHeartbeatResponseData response) {
     public void transitionToFenced() {
         resetEpoch();
         transitionTo(MemberState.FENCED);
+        assignmentReconciler.startLost();
```
the sequence here is not right I would say. Transitioning to FENCED currently means that the next poll will send out a HB with epoch 0, which should only be done when the `onPartitionsLost` callback completes. Wondering if an intermediate state like RELEASING_ASSIGNMENT would do. Similar to RECONCILING, but different because we do need to continue HB while reconciling, but not while releasing, because when fenced the member is not part of the group anymore until it rejoins. Both, reconciling and releasing would complete when the callbacks complete, but reconciling would transition to stable, vs. releasing that would transition to fenced (triggering the re-join). Makes sense?

philipnee replied:
I think this is actually fine, because the heartbeat won't be sent in the FENCED state. I see that completeLost() is invoked after the callback completes - there I think we can ask the member to rejoin by invoking tryJoin() (which is in my PR).

Technically - The member should try to rejoin after being fenced, but there's no guarantee to it. In this case - we rely on callback completion signal to ask the member to rejoin.  This is what i meant.

lianetm replied:
Uhm why do you say that "_the heartbeat won't be sent in the FENCED state_"? , that's not true in the current implementation. With the logic we have so far, the HB manager poll does generate a HB request on the interval in the FENCED state.

philipnee replied:
sorry, you are right, i kept forgetting about the changes - after reading the PR, due to the fact that we have to wait for the rebalance to complete to rejoin, maybe it is best to first move to the next state, i.e. UNJOINED (or something else is fine), and then send the heartbeat.

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java:
```diff
@@ -248,4 +261,28 @@ public void updateAssignment(ConsumerGroupHeartbeatResponseData.Assignment assig
+        }
+
+        assignmentReconciler.completeLost(lostPartitions);
+        transitionTo(MemberState.UNJOINED);
```
UNJOINED here seems not right, given that I see a difference between being FENCED (existing member out of a group, should join with epoch 0, and its current member ID) vs being UNJOINED (new member out of a group, should join with epoch 0 but without any member ID). It relates to my previous [comment](https://github.com/apache/kafka/pull/14357/files/a68a42d8548f0cbb53ac1bddaf013d8b7c190cb0..160462e8d515904fd0df9e8ffc42bdf56c8e5b76#r1361424454) about transitions when fenced. Thoughts?

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java:
```diff
@@ -248,4 +261,28 @@ public void updateAssignment(ConsumerGroupHeartbeatResponseData.Assignment assig
+
+        assignmentReconciler.completeLost(lostPartitions);
+        transitionTo(MemberState.UNJOINED);
+        // TODO: update state to signal the HeartbeatRequestManager to send an ACK heartbeat
```
uhm we do need to sort this out. One option I'm thinking about is I'm aiming at being able to have states that will easily indicate to the HB manager if it "shouldHeartbeatNow" (FENCED || RECONCILIATION_COMPLETED), making sure that we only transition to those when truly ready to HB (all callbacks completed). 
Right now the HB manager has a "shouldSendHeartbeat" but that's conceptually indicating more of a `shouldSkipHeartbeat` (if FAILED or NOT_IN_GROUP) so we need to be mindful about identifying both situations in the HB manager poll. 
Thoughts?

philipnee replied:
We just need to rejoin to reenable heartbeat right? We should'nt need to signal the HB manager explicitly.

lianetm replied:
you're right, not here, I meant it for the reconciliation complete which has the exact same TODO and we do need to sort that out. Thoughts?

lianetm replied:
Just for the record, after our last sync with @dajac we mentioned we could consider to send HB right away after the `onPartitionsLost` completes, without waiting for the interval, so the TODO applies here too.

lianetm replied:
Thinking more about this, wondering if this couple of new states would do:

- existing RECONCILING : starts when member gets new assignment and triggers callbacks. No HB.
- new RECONCILIATION_COMPLETE : starts when callbacks complete. Send HB on this state (and end the state when HB is sent out, transitioning to STABLE).
- existing FENCED (renamed to RELEASING_ASSIGNMENT): starts when member receives a fencing error, resets epoch to 0 and triggers callbacks. No HB. 
- new RELEASING_ASSIGNMENT_COMPLETE: starts when callback completes. Send HB on this state (and end state when HB is sent out, transitioning to UNJOINED)

Based on these, the HB manager could send HB based on the interval or if `shouldHeartBeatNow` true (state == RECONCILIATION_COMPLETE || RELEASING_ASSIGNMENT_COMPLETE)

kirktrue replied:
@dajac @lianetm @philipnee Please take a look at the [state machine diagram](https://cwiki.apache.org/confluence/display/KAFKA/Consumer+rebalance#Consumerrebalance-RebalanceStateMachine) on the AK wiki.

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AssignmentReconciler.java:
```diff
@@ -0,0 +1,263 @@
+ * {@code AssignmentReconciler} performs the work of reconciling this consumer's partition assignment as directed
+ * by the consumer group coordinator. When the coordinator determines that a change to the partition ownership of
+ * the group is required, it will communicate with each consumer to relay its respective <em>target</em>
+ * assignment, that is, the set of partitions for which that consumer should now assume ownership. It is the then the
```
It is the then the ...

kirktrue replied:
Will fix. Thanks!

kirktrue replied:
Done.

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AssignmentReconciler.java:
```diff
@@ -0,0 +1,263 @@
+ *         {@link MembershipManager#completeReconcile(Set, Set, Optional)} method. This method will call
+ *         {@link #completeReconcile(Set, Set)} to remove the revoked partitions and add the assigned partitions in the
+ *         {@link SubscriptionState#assignFromSubscribed(Collection) current assignment} and then make a note to
+ *         send a {@link ConsumerGroupHeartbeatRequest} to the group coordinator on its next pass of
```
maybe just "and then rejoin by doing ####"

kirktrue replied:
Sorry, can you clarify where this comment should be placed? My understanding is that rejoining isn't necessary after reconciliation.

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AssignmentReconciler.java:
```diff
@@ -0,0 +1,263 @@
+    private final Logger log;
+    private final SubscriptionState subscriptions;
+    private final ConsumerMetadata metadata;
+    private final BlockingQueue<BackgroundEvent> backgroundEventQueue;
```
We have a class called errorEventHandler - maybe we should rename/refacotr it and use it here.

kirktrue replied:
It's been changed a bit in the forthcoming fetch request manager PR to be more generic sense we need to also pass back non-error-related events to the application thread.

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AssignmentReconciler.java:
```diff
@@ -0,0 +1,263 @@
+
+        for (TopicPartitions topicPartitions : assignment.topicPartitions()) {
+            Uuid topicId = topicPartitions.topicId();
+            String topicName = Objects.requireNonNull(topicIdToNameMap.get(topicId), () -> String.format("No topic name was found in the metadata for topic ID %s", topicId));
```
The metadata cache may not have the relevant metadata for newly assigned topics so this won't work. I think that we must request a metadata update for the newly assigned topics.

kirktrue replied:
☹️ Let me think about how that affects things here...

kirktrue replied:
For my own understanding, when will a consumer receive assignments for topics which he hasn't subscribed to a priori? Is this the case where the user is using regex-based subscriptions?

cc @lianetm @philipnee

kirktrue replied:
The client currently _assumes_ that it can start reconciling immediately upon receipt of the new assignment. In this case where we're missing the topic metadata, we will need to:

1. Catch the case where we're missing metadata
2. Request the metadata update
3. Stash the assignment for later
4. Correctly note when we get the metadata update
5. Restart the processing of the reconciliation

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AssignmentReconciler.java:
```diff
@@ -0,0 +1,263 @@
+        return partitions;
+    }
+
+    private Map<Uuid, String> createTopicIdToNameMapping() {
+        Map<Uuid, String> map = new HashMap<>();
+
+        for (Map.Entry<String, Uuid> entry : metadata.topicIds().entrySet())
+            map.put(entry.getValue(), entry.getKey());
+
+        return map;
+    }
```
It seems to be that it would be better to maintain this mapping in the metadata cache if we need it. What do you think? I can tackle this separately though.

kirktrue replied:
That seems like a good idea. Perhaps as a separate PR?

kirktrue replied:
@dajac I added the changes for the topic mapping to this PR _for now_. I will split it out, but wanted to leave it here in case we end up going a different direction that obviates the need for this.

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AssignmentReconciler.java:
```diff
@@ -0,0 +1,263 @@
+            return;
+        }
+
+        log.debug("Enqueuing event to invoke {} callbacks", ConsumerRebalanceListener.class.getSimpleName());
+        PartitionReconciliationStartedEvent event = new PartitionReconciliationStartedEvent(
+                partitionsToRevoke,
+                partitionsToAssign
+        );
+        backgroundEventQueue.add(event);
```
Using rebalance callbacks is pretty uncommon, I think. Therefore, in most of the case, we won't have one. To be precise, it will be set to `NoOpConsumerRebalanceListener`. In this case, we don't have to go back to the application thread to execute nothing. We should just bypass it here. Have you considered this?

kirktrue replied:
Agreed.

See the [other comment](https://github.com/apache/kafka/pull/14357/files#r1362032050).

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AssignmentReconciler.java:
```diff
@@ -0,0 +1,263 @@
+            return;
+        }
+
+        log.debug("Enqueuing event to invoke {} callbacks", ConsumerRebalanceListener.class.getSimpleName());
+        backgroundEventQueue.add(new PartitionLostStartedEvent(partitionsToLose));
```
We could also skip this one if we don't have it.

kirktrue replied:
Yes, we initially wanted to skip this rigamarole, if possible. It's not clear to me how we know if we have a "valid" listener or not. Let me noodle on it a bit.

kirktrue replied:
Per #14565, we will have a way to check this unambiguously.

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AssignmentReconciler.java:
```diff
@@ -0,0 +1,263 @@
+     * @param assignment Holds the {@link Assignment} which includes the target set of topics
+     * @return Set of partitions to revoke
+     */
+
```
nit: This empty line could be removed.

kirktrue replied:
I'll remove the superfluous newline. Thanks.

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AssignmentReconciler.java:
```diff
@@ -0,0 +1,263 @@
+     */
+
+    SortedSet<TopicPartition> getPartitionsToRevoke(Assignment assignment) {
+        SortedSet<TopicPartition> partitions = new TreeSet<>(new TopicPartitionComparator());
```
Is there a reason why we use a SortedSet?

philipnee replied:
i think this is from the original implementation - I also wasn't sure why sorted set was used.

kirktrue replied:
Partially due to logging, but yes, it was that way before.

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AssignmentReconciler.java:
```diff
@@ -0,0 +1,263 @@
+     * @param assignment {@link Optional} that holds the {@link Assignment} which includes the target set of topics
+     * @return Set of partitions to assign
+     */
+
```
nit: Unnecessary empty line.

kirktrue replied:
I'll remove the superfluous newline. Thanks.

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AssignmentReconciler.java:
```diff
@@ -0,0 +1,263 @@
+        Set<TopicPartition> newAssignment = new HashSet<>(subscriptions.assignedPartitions());
+        newAssignment.addAll(assignedPartitions);
+        newAssignment.removeAll(revokedPartitions);
+        subscriptions.assignFromSubscribed(newAssignment);
```
For my understanding, I suppose that the fetcher will start to fetch from the newly assigned partitions after this point. Is my understanding right?

kirktrue replied:
Yes, though not immediately. When the fetcher does its next pass and sees the new partitions, they will be included in the next set of outgoing fetch requests.

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerUtils.java:
```diff
@@ -221,4 +229,72 @@ else if (t instanceof KafkaException)
         }
     }
+
+    static void processRebalanceCallback(final EventHandler eventHandler,
```
For my understanding, why do we place those methods in ConsumerUtils?

kirktrue replied:
Convenience for testing, mostly. Do you have a suggestion for a preferred location?

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:
```diff
@@ -1294,7 +1294,7 @@ boolean generationUnchanged() {
     }
 
-    protected final Meter createMeter(Metrics metrics, String groupName, String baseName, String descriptiveName) {
+    static Meter createMeter(Metrics metrics, String groupName, String baseName, String descriptiveName) {
```
It is a bit weird to just expose it like this. `createMeter` does not mean much outside of the usages within this class and sub-class. We could perhaps find a better place for this one and also find a better name. What do you think?

kirktrue replied:
Agreed.

I've updated how this is done in the refactoring PR #14565.

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerRebalanceListenerInvoker.java:
```diff
@@ -0,0 +1,118 @@
+    ConsumerRebalanceListenerInvoker(LogContext logContext,
+                                     SubscriptionState subscriptions,
+                                     Time time,
+                                     ConsumerCoordinatorMetrics sensors) {
```
nit: `consumerCoordinatorMetrics`?

kirktrue replied:
I changed the instance variable name to `coordinatorMetrics` in the refactoring PR #14565. It's a little shorter but still more accurate than `sensors`.

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerRebalanceListenerInvoker.java:
```diff
@@ -0,0 +1,118 @@
+        this.sensors = sensors;
+    }
+
+    Exception invokePartitionsAssigned(final SortedSet<TopicPartition> assignedPartitions) {
```
Those methods seem to be a one-to-one copy of the originals. Is this correct?

kirktrue replied:
Yes, correct.

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessor.java:
```diff
@@ -155,8 +164,18 @@ private boolean processValidatePositionsEvent() {
     private boolean process(final TopicMetadataApplicationEvent event) {
         final CompletableFuture<Map<String, List<PartitionInfo>>> future =
-            this.requestManagers.topicMetadataRequestManager.requestTopicMetadata(Optional.of(event.topic()));
+                this.requestManagers.topicMetadataRequestManager.requestTopicMetadata(Optional.of(event.topic()));
```
Let's revert this change. It is not related.

kirktrue replied:
Will do.

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/PartitionLostCompleteEvent.java:
```diff
@@ -0,0 +1,75 @@
+        return result;
+    }
+
+    @Override
+    protected String toStringBase() {
+        return super.toStringBase() + ", lostPartitions=" + lostPartitions + ", error=" + error;
+    }
+
+    @Override
+    public String toString() {
+        return getClass().getSimpleName() + "{" +
+                toStringBase() +
+                '}';
+    }
```
This seems to be a quite complicated way to implement `toString`. When I read it, I cannot even tell how it will look like... Why don't we simply implement toString like we do in all classes in the code base?

philipnee replied:
I think this is from Jun's comment in one of the commit that we should try to reuse the toString from its inherited class.

kirktrue replied:
Yes. We used to have it the bog standard way, but there was concern about its maintainability, so we adopted this approach 🤷‍♂️

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/PartitionLostCompleteEvent.java:
```diff
@@ -0,0 +1,75 @@
+import java.util.Optional;
+import java.util.SortedSet;
+
+public class PartitionLostCompleteEvent extends ApplicationEvent {
```
nit: Let's add some javadoc.

kirktrue replied:
Will do.

kirktrue replied:
Done.

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/PartitionLostStartedEvent.java:
```diff
@@ -0,0 +1,66 @@
+import java.util.Collections;
+import java.util.SortedSet;
+
+public class PartitionLostStartedEvent extends BackgroundEvent {
```
nit: Let's add some javadoc.

kirktrue replied:
Will do.

kirktrue replied:
Done.

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java:
```diff
@@ -248,4 +261,28 @@ public void updateAssignment(ConsumerGroupHeartbeatResponseData.Assignment assig
+
+        assignmentReconciler.completeReconcile(revokedPartitions, assignedPartitions);
+        transitionTo(MemberState.STABLE);
+        // TODO: update state to signal the HeartbeatRequestManager to send an ACK heartbeat
```
We need to sort this out. Wondering if new states/transitions, used from the HB poll would do...as described in my other comment [here](https://github.com/apache/kafka/pull/14357/files#r1361441480)?

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/RequestManagers.java:
```diff
@@ -32,7 +32,7 @@ public class RequestManagers {
     public final Optional<CoordinatorRequestManager> coordinatorRequestManager;
     public final Optional<CommitRequestManager> commitRequestManager;
-    private final Optional<HeartbeatRequestManager> heartbeatRequestManager;
+    public final Optional<HeartbeatRequestManager> heartbeatRequestManager;
```
I will revert this change since we didn't end up using the heart beat request manager elsewhere.

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/PartitionReconciliationCompleteEvent.java:
```diff
@@ -0,0 +1,89 @@
+import java.util.Optional;
+import java.util.SortedSet;
+
+public class PartitionReconciliationCompleteEvent extends ApplicationEvent {
```
I'll add JavaDoc here, also.

kirktrue replied:
Done.

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/PartitionReconciliationStartedEvent.java:
```diff
@@ -0,0 +1,77 @@
+import java.util.Collections;
+import java.util.SortedSet;
+
+public class PartitionReconciliationStartedEvent extends BackgroundEvent {
```
And some add JavaDoc here, too.

kirktrue replied:
Done.

kirktrue committed:
Updates to assignment reconciliation based on feedback

* Adds JavaDoc
* Fixes typos in JavaDoc
* Removes unnecessary whitespace
* Reverts unrelated changes

kirktrue reviewed (commented):
Marked completed tasks as 'done'.

kirktrue committed:
Added markPartitionsPendingRevocation and added/cleaned up comments

kirktrue committed:
Updates for topic mapping and sketching out handing of missing topic names in the local metadata cache

kirktrue committed:
Merge remote-tracking branch 'upstream/trunk' into KAFKA-15276-partition-reconciliation
